packages/flutter_tools/lib/src/asset.dart

```diff
@@ -1026,39 +1045,47 @@ class _AssetDirectoryCache {
 
   final FileSystem _fileSystem;
   final List<String> _excluded;
-  final Map<String, Map<String, List<String>>> _cache = <String, Map<String, List<String>>>{};
+  final Map<String, List<String>> _cache = <String, List<String>>{};
 
   List<String> variantsFor(String assetPath) {
-    final String assetName = _fileSystem.path.basename(assetPath);
     final String directory = _fileSystem.path.dirname(assetPath);
 
     if (!_fileSystem.directory(directory).existsSync()) {
       return const <String>[];
     }
 
-    if (_cache[directory] == null) {
-      final List<String> paths = <String>[];
-      for (final FileSystemEntity entity in _fileSystem.directory(directory).listSync(recursive: true)) {
-        final String path = entity.path;
-        if (_fileSystem.isFileSync(path)
-          && assetPath != path
-          && !_excluded.any((String exclude) => _fileSystem.path.isWithin(exclude, path))) {
-          paths.add(path);
-        }
-      }
-
-      final Map<String, List<String>> variants = <String, List<String>>{};
-      for (final String path in paths) {
-        final String variantName = _fileSystem.path.basename(path);
-        if (directory == _fileSystem.path.dirname(path)) {
-          continue;
-        }
-        variants[variantName] ??= <String>[];
-        variants[variantName]!.add(path);
-      }
-      _cache[directory] = variants;
+    if (_cache.containsKey(assetPath)) {
+      return _cache[assetPath]!;
     }
 
-    return _cache[directory]![assetName] ?? const <String>[];
+    final List<FileSystemEntity> entitiesInDirectory = _fileSystem.directory(directory).listSync();
+
+    final List<String> pathsOfPossibleVariants = <String>[
+      // It's possible that the user specifies only explicit variants (e.g. .../1x/asset.png),
+      // so there does not necessarily need to be a file at the given path.
+      if (_fileSystem.file(assetPath).existsSync()) assetPath,
+      ...entitiesInDirectory
+        .whereType<Directory>()
+        .where((Directory dir) => _assetVariantDirectoryRegExp.hasMatch(dir.basename))
+        .map((Directory dir) => dir.listSync())
+        .flatten()
```

Comment:
I don't think this extension method is carrying its weight. You don't need to `toList` it, since whereType operates on an iterable as well

Reply:
Also https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-extension

Reply:
It doesn't just call `toList`, it more importantly flattens the list of _lists of_ directories into a plain list of directories.

When I was originally looking at this, I was looking for a Dart analogue to `flatMap` ([JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap), [Java](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#flatMap-java.util.function.Function-)). I just now, however, realized that [`expand`](https://api.flutter.dev/flutter/dart-core/Iterable/expand.html) is what I was looking for the whole time. I replaced this extension method with that.